### PR TITLE
Core - REST Catalog call /v1/config route

### PR DESCRIPTION
**Pull request by kbendick**
Adds the initial call to `/v1/config` and merges the server-side configuration with the client configuration.

Also adds in an object mapper specific to the RESTCatalog, so that the object mapper in `JsonUtils` doesn't need to be modified.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/RESTObjectMapper.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      synchronized (RESTObjectMapper.class) {
+        if (!isInitialized) {
+          MAPPER.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+          RESTSerializers.registerAll(MAPPER);
+          isInitialized = true;
+        }
+      }
+    }
+
+    return MAPPER;
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/rest/RESTObjectMapper.java:46]: We can move this to another PR if we'd like.

We need an object mapper that is configured with the necessary visibility modifiers.

I made it lazy like this so that people who don't need this don't wind up loading two object mappers, but that could be overkill.

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * RESTCatalog.
+ */
+public class HTTPClientFactory implements RESTClientFactory {
+
+  @Override
+  public RESTClient apply(Map<String, String> properties) {
+    Preconditions.checkArgument(properties != null, "Invalid initial configuration: null");
+    Preconditions.checkArgument(properties.containsKey(CatalogProperties.URI), "REST Catalog server URI is required");
+    Preconditions.checkArgument(properties.containsKey(RESTCatalogProperties.AUTH_TOKEN),
+        "Invalid http client configuration. Missing authentication token");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java:39]: Is this required? I think we should use it if it's there, but we don't need to force other people into using an auth token.

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * Takes in the full configuration for the {@link RESTCatalog}, which should already have
+ * called the server's initial configuration route.
+ * Using the merged configuration, an instance of {@link RESTClient} is obtained that can be used with the
+ * RESTCatalog.
+ */
+public class HTTPClientFactory implements RESTClientFactory {
+
+  @Override
+  public RESTClient apply(Map<String, String> properties) {
+    Preconditions.checkArgument(properties != null, "Invalid initial configuration: null");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java:36]: Remove "initial"? I don't think that makes sense if this is called with a merged configuration after fetching server-side config.

File: core/src/main/java/org/apache/iceberg/rest/RESTClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.rest;
+
+import java.util.Map;
+import java.util.function.Function;
+
+public interface RESTClientFactory extends Function<Map<String, String>, RESTClient> {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTClientFactory.java:25]: What is the purpose of this interface? I don't really see how this would be used.

File: core/src/main/java/org/apache/iceberg/rest/RESTObjectMapper.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+package org.apache.iceberg.rest;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class RESTObjectMapper {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTObjectMapper.java:27]: This leaks Jackson, so it should not be public. Can you make the class and the `mapper` method package-private?

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    Preconditions.checkArgument(properties.containsKey(CatalogProperties.URI), "REST Catalog server URI is required");
+    Preconditions.checkArgument(properties.containsKey(RESTCatalogProperties.AUTH_TOKEN),
+        "Invalid http client configuration. Missing authentication token");
+
+    String baseURI = properties.get(CatalogProperties.URI);
+
+    return HTTPClient.builder()
+        .withBearerAuth(properties.get(RESTCatalogProperties.AUTH_TOKEN))
+        .uri(baseURI)
+        .mapper(RESTObjectMapper.mapper())
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java:46]: Can you remove `mapper` from the client builder API? I don't think it is something that we need to expose and it leaks Jackson in the API.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 7]
```diff
@@ -69,18 +75,66 @@
...
+    Preconditions.checkArgument(initialProps.containsKey(RESTCatalogProperties.AUTH_TOKEN),
+        "Invalid http client configuration. Missing authentication token");
+
+    String baseURI = initialProps.get(CatalogProperties.URI).trim();
+
+    HTTPClient singleUseClient = HTTPClient.builder()
+        .uri(baseURI)
+        .mapper(RESTObjectMapper.mapper())
+        .withBearerAuth(initialProps.get(RESTCatalogProperties.AUTH_TOKEN).trim())
+        .build();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:124]: Why does this duplicate the logic from `HTTPClientFactory`? The intent of having the client builder was to be able to use it in both places.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 8]
```diff
@@ -69,18 +75,66 @@
...
+    initializeInternal(name, mergedProps);
+  }
+
+  /**
+   * Performs the normal initialization steps after calling the config route and merging the initial
+   * configuration with any configuration provided by the server.
+   * <p>
+   * See {@link RESTCatalogConfigResponse} for details on merging the initial client and server configuration.
+   *
+   * @param name  Name of this RESTCatalog
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:98]: I think that arg names should match the variable names. Could you make the case match?

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 9]
```diff
@@ -69,18 +75,66 @@
...
+   * @param props Properties to use, merged with any server-provided configuration.
+   */
+  @VisibleForTesting
+  void initializeInternal(String name, Map<String, String> props) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:102]: I think we can avoid adding both of these `@VisibleForTesting` methods. To test, I think you can either update `RESTCatalogAdaptor` to accept a config map to return, or just pass in a different `RESTClient` when creating a test catalog. It could be something roughly like this:

```java
  @Test
  public void testConfigRoute() {
    RESTClient testClient = new RESTClient() {
      @Override
      public void head(String path, Consumer<ErrorResponse> errorHandler) {
        throw new UnsupportedOperationException("Should not be called for testConfigRoute");
      }

      @Override
      public <T extends RESTResponse> T delete(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
        throw new UnsupportedOperationException("Should not be called for testConfigRoute");
      }

      @Override
      public <T extends RESTResponse> T get(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
      }

      @Override
      public <T extends RESTResponse> T post(String path, RESTRequest body, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
        throw new UnsupportedOperationException("Should not be called for testConfigRoute");
      }

      @Override
      public void close() {
      }
    };

    RESTCatalog restCatalog = new RESTCatalog((config) -> testClient);

    restCatalog.setConf(new Configuration());
    restCatalog.initialize("prod", ImmutableMap.of());
    
    Assert.assertNotNull("Should have config key foo", restCatalog.properties().get("foo"));
  }
```

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java (hunk 2 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java:39]: I was wondering the same thing. I had originally had it as optional.

I'll change it so that it is only applied if the user provides it.

File: core/src/main/java/org/apache/iceberg/rest/RESTObjectMapper.java (hunk 5 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTObjectMapper.java:27]: Yes I can.

File: core/src/main/java/org/apache/iceberg/rest/RESTClientFactory.java (hunk 4 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTClientFactory.java:25]: Originally, this had more recognizable benefit. I decided to keep it in so that the hierarchy of `REST*` being the interface and then `HTTP*` being the implementing class, like `RESTClient` <-> `HTTPClient`, would be kept.

However, since this is a plain functional interface, I agree it doesn't provide much benefit. Will remove.

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java (hunk 3 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java:36]: Agreed. Originally this handled the merging, but since the client builder / factory is used in other places, I removed that.

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java (hunk 6 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java:43]: Sure. Now that we have one, it's not really needed.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 8 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:98]: Maybe this comment is dated, but if you're referring to the parameter names in the javadoc comment, they do match.

If you're referring to making the parameter named `properties`, I think that will cause a shadowed variable error in error prone since this isn't a constructor. Let me double check.

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java (hunk 2 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java]: Updated to use the auth token only if present.

File: core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java (hunk 3 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/HTTPClientFactory.java]: Removed.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 7 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java]: Yeah. I noticed that after posting it. Originally what I was working with for the `HTTPClientFactory` had some stuff that made it not possible to share the code. But that's been fixed so things have been deduplicated!

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 8 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:98]: Oh, looks like my bad. Sorry about that, not sure why I didn't see `name`.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 9 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:102]: So this did still require 2 `VisibleForTesting` functions, but one of them is just an accessor for `properties()`. I was able to remove the annotation for the function for merging in server config, as well as the usage of mockito.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 8 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:98]: Ah no worries. It happens.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 10]
```diff
@@ -72,4 +78,62 @@ protected boolean supportsNamespaceProperties() {
...
+  public void testInitializeWithBadArguments() {
+    AssertHelpers.assertThrows("Configuration passed to initialize cannot be null",
+        IllegalArgumentException.class,
+        "Invalid configuration: null",
+        () -> catalog().initialize("prod", null));
+
+    AssertHelpers.assertThrows("Configuration passed to initialize must have uri",
+        IllegalArgumentException.class,
+        "REST Catalog server URI is required",
+        () -> catalog().initialize("prod", ImmutableMap.of()));
```
Comment (kbendick) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:137]: Now that we've refactored to not have such a large surface area in `VisibleForTesting`, I'm going to move these tests over to a test class for `TestHTTPClientFactory`.

They're failing now because I updated `RESTCatalog`'s fetch server configuration function to use the passed in adaptor (via `clientBuilder`).

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 11]
```diff
@@ -55,7 +61,7 @@ public void createCatalog() {
 
     this.restCatalog = new RESTCatalog((config) -> adaptor);
     restCatalog.setConf(conf);
-    restCatalog.initialize("prod", ImmutableMap.of());
+    restCatalog.initializeInternal("prod", ImmutableMap.of());
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:64]: Why does this call `initializeInternal` still? I think that normal initialization should be used for testing.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 12]
```diff
@@ -72,4 +78,63 @@ protected boolean supportsNamespaceProperties() {
...
+      }
+    };
+
+    RESTCatalog restCat = new RESTCatalog((config) -> testClient);
+    Map<String, String> initialConfig = ImmutableMap.of("uri", "http://localhost:8080");
+
+    restCat.setConf(new Configuration());
+    restCat.initialize("prod", initialConfig);
+
+    Assert.assertNotNull("Should have sever-provided config keys", restCat.properties().get("cache-enabled"));
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:124]: Typo: `sever` -> `server`.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 13]
```diff
@@ -72,4 +78,63 @@ protected boolean supportsNamespaceProperties() {
...
+                                               Consumer<ErrorResponse> errorHandler) {
+        throw new UnsupportedOperationException("Should not be called for testConfigRoute");
+      }
+
+      @Override
+      public <T extends RESTResponse> T get(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+        return (T) RESTCatalogConfigResponse
+            .builder()
+            .withDefaults(ImmutableMap.of())
+            .withOverrides(ImmutableMap.of("cache-enabled", "false"))
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:103]: Right now, this test uses disjoint property sets, but I'd like to see it validate the override behavior. I realize that's tested in the config response, but at least a basic check would be good here.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 14]
```diff
@@ -69,18 +75,42 @@
...
-    String ioImpl = props.get(CatalogProperties.FILE_IO_IMPL);
-    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), props, conf);
+    this.catalogName = name;
+    this.paths = ResourcePaths.forCatalogProperties(properties);
+    String ioImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
+    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), properties, conf);
+  }
+
+  @VisibleForTesting
+  Map<String, String> properties() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:112]: This could just be public, too. I think it's reasonable for the catalog to expose its config properties.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 15]
```diff
@@ -69,18 +75,42 @@
...
     this.client = clientBuilder.apply(props);
-    this.catalogName = name;
     this.properties = ImmutableMap.copyOf(props);
-    this.paths = ResourcePaths.forCatalogProperties(props);
-    String ioImpl = props.get(CatalogProperties.FILE_IO_IMPL);
-    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), props, conf);
+    this.catalogName = name;
+    this.paths = ResourcePaths.forCatalogProperties(properties);
+    String ioImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
+    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), properties, conf);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:108]: It looks like these lines don't need to change at all. The current set of changes is to reorder some of them (move catalogName assignment) and rename `props` to `properties`. I think we can avoid making at least some of these changes. But I would merge the methods together first, since that will affect this.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 16]
```diff
@@ -405,4 +435,23 @@ private String fullTableName(TableIdentifier ident) {
...
+    try {
+      RESTCatalogConfigResponse configResponse = singleUseClient
+          .get(ResourcePaths.config(), RESTCatalogConfigResponse.class, ErrorHandlers.defaultErrorHandler());
+      configResponse.validate();
+      return configResponse.merge(props);
+    } finally {
+      try {
+        singleUseClient.close();
+      } catch (IOException e) {
+        LOG.error("Failed to close http client used for getting catalog configuration. Possible resource leak.", e);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:453]: `http` should be all caps.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 17]
```diff
@@ -405,4 +435,23 @@ private String fullTableName(TableIdentifier ident) {
 
     return Pair.of(tableClient, tableIO);
   }
+
+  private Map<String, String> fetchAndMergeWithServerConf(Map<String, String> props) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:439]: I think this would be cleaner if it just fetched the config and returned the response. That way it does one only thing. The only call that would be moved outside is the `merge` call, which probably shouldn't be in the `try` block anyway.

Then the caller would look like this:

```java
    ConfigResponse config = fetchConfig(properties);
    Map<String, String> props = config.merge(properties);
```

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 10 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java]: Actually these didn't need to be moved. I will consider it if need be.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 13 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:114]: Ok. I wasn't sure how much testing to add here because there are tests over there, but I agree it would be a good idea to have a bit more extensive testing here.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 14 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:112]: Ok. Given that the properties can change (mostly from the config call), I wasn't sure if we wanted to expose it. But that seems fair.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 18]
```diff
@@ -69,18 +75,42 @@
...
+   * Performs the normal initialization steps after calling the config route and merging the initial
+   * configuration with any configuration provided by the server.
+   * <p>
+   * See {@link RESTCatalogConfigResponse} for details on merging the initial client and server configuration.
+   *
+   * @param name  Name of this RESTCatalog
+   * @param props Properties to use, merged with any server-provided configuration.
+   */
+  @VisibleForTesting
+  void initializeInternal(String name, Map<String, String> props) {
```
[the remark this answers is not in the document]
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java]: Yeah this was removed.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 17 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:444]: Yeah I had already moved it to the `try` block, but I will move the merging out so that this function has more of a single purpose.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 19]
```diff
@@ -42,20 +50,29 @@ public void createCatalog() {
...
-        JdbcCatalog.PROPERTY_PREFIX + "password", "password");
-    backendCatalog.initialize("backend", backendCatalogProperties);
+    this.restCatalog = new RESTCatalog((config) -> {
+      JdbcCatalog backendCatalog = new JdbcCatalog();
+      backendCatalog.setConf(conf);
+      Map<String, String> backendCatalogProperties = ImmutableMap.of(
+          CatalogProperties.WAREHOUSE_LOCATION, warehouse.getAbsolutePath(),
+          CatalogProperties.URI, "jdbc:sqlite:file::memory:?ic" + UUID.randomUUID().toString().replace("-", ""),
+          JdbcCatalog.PROPERTY_PREFIX + "username", "user",
+          JdbcCatalog.PROPERTY_PREFIX + "password", "password");
```
Comment (kbendick) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:60]: The `backendCatalog` creation needed to be moved into the constructor so that the single use connection that is opened and then closed to fetch the server config doesn't permanently close the underlying JDBC catalog connection.

This way, the lifecycle of the connection to the JDBC catalog mirrors that of the HTTP client.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 16 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java]: Updated.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 17 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java]: Updated.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalogProperties.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+ */
+public class RESTCatalogProperties {
+
+  private RESTCatalogProperties() {
+  }
+
+  /**
+   * A Bearer authorization token which will be used to authenticate requests with the server.
+   */
+  public static final String AUTH_TOKEN = "token";
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalogProperties.java:34]: Do we want this to start with `rest.`, similar to the way that `jdbc` catalog prefaces its configs?

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 11 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java]: No longer calls `initializeInternal`.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 15 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java]: Updated so that these don't change.

File: core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java [hunk 21]
```diff
@@ -89,6 +90,11 @@ public CommitFailedException wrapped() {
     }
   }
 
+  public static RESTCatalogConfigResponse fetchServerConfig() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java:93]: What is the purpose of updating `CatalogHandlers`? Since this doesn't actually do translation to the `Catalog` API, it doesn't seem like something we should provide a helper for. Plus, this doesn't give the caller any opportunity to customize the config.

File: core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 22]
```diff
@@ -154,8 +154,7 @@ public <T extends RESTResponse> T handleRequest(Route route, Map<String, String>
                                                   Object body, Class<T> responseType) {
     switch (route) {
       case CONFIG:
-        // TODO: use the correct response object
-        return castResponse(responseType, ImmutableMap.of());
+        return castResponse(responseType, CatalogHandlers.fetchServerConfig());
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:157]: I commented above that it probably isn't very helpful to include `fetchServerConfig()` in `CatalogHandlers`. I would recommend creating the response directly here. You could also include a default config map for testing purposes if you wanted, although I think you're testing it different way so you may not need it.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 19 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:60]: I'm a bit concerned about this because it creates an entirely new JDBC catalog. It works for tests now, but if we were to send slightly different config back when loading a table, the catalog would create a new connection and wipe the DB.

Could this be solved instead by skipping `catalog.close()` in the `RESTCatalogAdaptor.close` method? We could simply close the `backendCatalog` in an `@AfterEach` method.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 23]
```diff
@@ -72,4 +90,72 @@ protected boolean supportsNamespaceProperties() {
   protected boolean supportsServerSideRetry() {
     return true;
   }
+
+  /* RESTCatalog specific tests */
+
+  @Test
+  public void testConfigRoute() throws IOException {
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:97]: These tests look good to me.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 12 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java]: Fixed.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 13 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java]: Added more testing for defaults and overrides being applied. Between that and the tests actually in the `RESTCatalogConfigResponse`, things should be adequately tested.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 19 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:60]: Yeah that would be fine. I had it that way at one point in time (but never pushed it), so let me get it back that way again.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalogProperties.java (hunk 20 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalogProperties.java:34]: Yeah I don't think so either. Should be easy to add if need be.

File: core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java (hunk 21 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java:93]: The tests fail if there's not something in place of this call, since we call `initialize` and not the older `initializeInternal`.

So something needs to be returned here.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 24]
```diff
@@ -69,12 +75,31 @@ public class RESTCatalog implements Catalog, SupportsNamespaces, Configurable<Co
...
+  /**
+   * Performs the normal initialization steps after calling the config route and merging the initial
+   * configuration with any configuration provided by the server.
+   * <p>
+   * See {@link RESTCatalogConfigResponse} for details on merging the initial client and server configuration.
+   *
+   * @param name  Name of this RESTCatalog
+   * @param props Properties to use, merged with any server-provided configuration.
+   */
+  private void initializePostConfig(String name, Map<String, String> props) {
```
[the remark this answers is not in the document]
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:102]: My main reasoning for making it a separate function was so that the non-merged properties didn't get reused accidentally. I can merge it into one function if we prefer.

File: core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java (hunk 22 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:157]: Sure I can inline it here. That makes sense.

And since the server configuration being applied depends on what configuration is passed in, I chose to forgo it for now since it's tested elsewhere.

File: core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 25]
```diff
@@ -281,9 +281,9 @@ public void head(String path, Consumer<ErrorResponse> errorHandler) {
 
   @Override
   public void close() throws IOException {
-    if (catalog instanceof Closeable) {
-      ((Closeable) catalog).close();
-    }
+    // The calling test is responsible for closing the underlying catalog backing this REST catalog
+    // so that the underlying backend catalog is not closed and reopened during the REST catalog's
+    // initialize method when fetching the server configuration.
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:286]: Looks good.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 26]
```diff
@@ -55,7 +64,19 @@ public void createCatalog() {
...
+  }
+
+  @AfterEach
+  public void closeCatalog() throws IOException {
+    if (restCatalog != null) {
+      restCatalog.close();
+    }
+
+    if (backendCatalog != null) {
+      this.backendCatalog.close();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:78]: Nit: unnecessary `this.` prefix. We only require it when assigning to an instance field.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java [hunk 27]
```diff
@@ -69,18 +75,28 @@ public class RESTCatalog implements Catalog, SupportsNamespaces, Configurable<Co
...
-    this.client = clientBuilder.apply(props);
+    RESTCatalogConfigResponse config = fetchConfig(props);
+    Map<String, String> mergedProps = config.merge(props);
+    this.client = clientBuilder.apply(mergedProps);
     this.catalogName = name;
-    this.properties = ImmutableMap.copyOf(props);
-    this.paths = ResourcePaths.forCatalogProperties(props);
-    String ioImpl = props.get(CatalogProperties.FILE_IO_IMPL);
-    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), props, conf);
+    this.properties = ImmutableMap.copyOf(mergedProps);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java:92]: Nit: `RESTCatalogConfigResponse.merge` returns an `ImmutableMap` already.

**Review by rdblue — APPROVED** [review]
+1 once checkstyle is fixed. Here's the error:

```
[ERROR] /home/runner/work/iceberg/iceberg/core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:22:8: Unused import - java.io.Closeable. [UnusedImports]
```

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java (hunk 26 above)
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:78]: Good catch.

File: core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java (hunk 27 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java]: Removed.

**Comment by kbendick** [comment]
> +1 once checkstyle is fixed. Here's the error:
> 
> ```
> [ERROR] /home/runner/work/iceberg/iceberg/core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:22:8: Unused import - java.io.Closeable. [UnusedImports]
> ```

This has been removed. Once CI completes this should be good to go.

**Comment by rdblue** [comment]
Thanks, @kbendick!
